### PR TITLE
Building Event service [IA-206]

### PR DESCRIPTION
**Pull request by AlexWeinstein92**

### CODE REVIEW

File: common-types/src/main/scala/com/improving/app/common/errors/Validation.scala [hunk 1]
```diff
@@ -80,37 +81,16 @@ object Validation {
...
-
-  val storeIdValidator: Validator[StoreId] = storeId => {
-    if (storeId.id.isEmpty) {
-      Some(ValidationError("Store Id is empty"))
-    } else {
-      None
+  def endBeforeStartValidator(startTime: Option[Timestamp]): Validator[Timestamp] = endTime =>
+    startTime match {
+      case Some(startTime) =>
+        if (endTime.seconds <= startTime.seconds) {
```
Comment (dag-yoppworks) [inline common-types/src/main/scala/com/improving/app/common/errors/Validation.scala:87]: Are there issues with 0-length events we foresee?

File: event/src/main/scala/com/improving/app/event/domain/Event.scala [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+          )
+        )
+      )
+    )
+    info match {
+      case Left(e: EditableEventInfo) =>
+        val validationErrorsOpt = draftTransitionEventInfoValidator(e)
+        if (validationErrorsOpt.isEmpty) {
+          Right(
+            EventEventResponse(
```
Comment (dag-yoppworks) [inline event/src/main/scala/com/improving/app/event/domain/Event.scala:488]: EventEventResponse :\ I wonder if we would do well to find an alternate name for Event to avoid collision with actor model internal nomenclature.

Comment (dag-yoppworks) [reply event/src/main/scala/com/improving/app/event/domain/Event.scala:488]: https://www.thesaurus.com/browse/event I think "Affair" is the closest in meaning but "Event" is much more idiomatic in the domain, I think. Oh well.

**Review by dag-yoppworks — COMMENTED** [review]
Looks good so far.

File: common-types/src/main/scala/com/improving/app/common/errors/Validation.scala (hunk 1 above)
Comment (AlexWeinstein92) [reply common-types/src/main/scala/com/improving/app/common/errors/Validation.scala:87]: I think those should be validated by gateway

File: event/src/main/scala/com/improving/app/event/api/EventServiceImpl.scala [hunk 3]
```diff
@@ -141,17 +141,9 @@ class EventServiceImpl()(implicit val system: ActorSystem[_]) extends EventServi
...
-                Some(_ @EventInfo(_, _, _, _, Some(expectedStart), _, _, _, _)),
-                _,
-                _
-              ),
-              _
-            )
+            EventEventResponse(response @ EventScheduled(_, _, _, _), _)
           ) =>
-        startStartTimer(in, expectedStart, id)
+        // startStartTimer(in, expectedStart, id)
```
Comment (dag-yoppworks) [inline event/src/main/scala/com/improving/app/event/api/EventServiceImpl.scala:146]: Commented out code

File: event/src/main/scala/com/improving/app/event/api/EventServiceImpl.scala [hunk 4]
```diff
@@ -162,9 +154,9 @@ class EventServiceImpl()(implicit val system: ActorSystem[_]) extends EventServi
    */
   override def cancelEvent(in: CancelEvent): Future[EventCancelled] = handleCommand(
     in,
-    { case StatusReply.Success(EventEventResponse(response @ EventCancelled(Some(id), _, _), _)) =>
-      cancelStartTimer(id)
-      cancelEndTimer(id)
+    { case StatusReply.Success(EventEventResponse(response @ EventCancelled(_, _, _), _)) =>
+      // cancelStartTimer(id)
+      // cancelEndTimer(id)
```
Comment (dag-yoppworks) [inline event/src/main/scala/com/improving/app/event/api/EventServiceImpl.scala:159]: more commented out code

File: event/src/main/scala/com/improving/app/event/api/EventServiceImpl.scala [hunk 5]
```diff
@@ -176,18 +168,10 @@ class EventServiceImpl()(implicit val system: ActorSystem[_]) extends EventServi
...
-                _
-              ),
-              _
-            )
+            EventEventResponse(response @ EventRescheduled(_, _, _, _), _)
           ) =>
-        cancelStartTimer(id)
-        startStartTimer(in, expectedStart, id)
+        // cancelStartTimer(id)
+        // startStartTimer(in, expectedStart, id)
```
Comment (dag-yoppworks) [inline event/src/main/scala/com/improving/app/event/api/EventServiceImpl.scala:174]: more commented out code

File: event/src/main/scala/com/improving/app/event/api/EventServiceImpl.scala [hunk 6]
```diff
@@ -200,13 +184,13 @@ class EventServiceImpl()(implicit val system: ActorSystem[_]) extends EventServi
...
             EventEventResponse(
-              response @ EventDelayed(Some(id), Some(_ @EventInfo(_, _, _, _, Some(expectedStart), _, _, _, _)), _, _),
+              response @ EventDelayed(_, _, _, _),
               _
             )
           ) =>
-        if (cancellableStartTimers.contains(id.id)) cancelStartTimer(id)
-        if (cancellableEndTimers.contains(id.id)) cancelEndTimer(id)
-        startStartTimer(in, expectedStart, id)
+        // if (cancellableStartTimers.contains(id.id)) cancelStartTimer(id)
```
Comment (dag-yoppworks) [inline event/src/main/scala/com/improving/app/event/api/EventServiceImpl.scala:191]: more commented out code, and `if` with no braces should be avoided

File: event/src/main/scala/com/improving/app/event/api/EventServiceImpl.scala [hunk 7]
```diff
@@ -220,12 +204,12 @@ class EventServiceImpl()(implicit val system: ActorSystem[_]) extends EventServi
...
             EventEventResponse(
-              response @ EventStarted(Some(id), Some(_ @EventInfo(_, _, _, _, _, Some(expectedEnd), _, _, _)), _, _),
+              response @ EventStarted(_, _, _, _),
               _
             )
           ) =>
-        cancelStartTimer(id)
-        startEndTimer(in, expectedEnd, id)
+        // cancelStartTimer(id)
+        // startEndTimer(in, expectedEnd, id)
```
Comment (dag-yoppworks) [inline event/src/main/scala/com/improving/app/event/api/EventServiceImpl.scala:212]: more commented out code

File: event/src/main/scala/com/improving/app/event/domain/Event.scala [hunk 8]
```diff
@@ -139,34 +132,57 @@ object Event extends StrictLogging {
...
+                            StateError(s"${command.request.productPrefix} command cannot be used on a scheduled Event")
                           )
                       }
                     case InProgressEventState(info, meta) =>
                       command.request match {
                         case delayEventCommand: DelayEvent =>
                           delayEvent(Right(info), meta, delayEventCommand)
                         case endEventCommand: EndEvent =>
-                          endEvent(meta, endEventCommand)
+                          if (Instant.now().getEpochSecond > info.getExpectedEnd.seconds)
```
Comment (dag-yoppworks) [inline event/src/main/scala/com/improving/app/event/domain/Event.scala:160]: maybe should be `>=` ?

File: event/src/main/scala/com/improving/app/event/domain/Event.scala [hunk 9]
```diff
@@ -139,34 +132,57 @@ object Event extends StrictLogging {
                   x match {
                     case ScheduledEventState(info, meta) =>
                       command.request match {
-                        case startEventCommand: StartEvent => startEvent(info, meta, startEventCommand)
-                        case delayEventCommand: DelayEvent => delayEvent(Right(info), meta, delayEventCommand)
+                        case startEventCommand: StartEvent =>
+                          if (Instant.now().getEpochSecond > info.getExpectedStart.seconds)
```
Comment (dag-yoppworks) [inline event/src/main/scala/com/improving/app/event/domain/Event.scala:136]: maybe should be `>=`?

Comment (dag-yoppworks) [reply event/src/main/scala/com/improving/app/event/domain/Event.scala:136]: Maybe a little slop even on the early side would be nice, to allow an event to start a few minutes early if the user requests it

File: event/src/main/scala/com/improving/app/event/domain/Event.scala (hunk 8 above)
Comment (dag-yoppworks) [reply event/src/main/scala/com/improving/app/event/domain/Event.scala:160]: And similarly on the end, ending early is a thing that happens sometimes and it might be nice to be able to do that without a big rescheduling workflow

**Review by dag-yoppworks — COMMENTED** [review]
Tests look okay. I had a couple questions about boundary conditions for starting and ending events.

**Comment by AlexWeinstein92** [comment]
> Tests look okay. I had a couple questions about boundary conditions for starting and ending events.

I was originally putting in timers for automatically starting and ending events, thinking that was desired for the demo. But Reid said that debugging things like that can be a nightmare so I should just focus on letting users start and end events.

For Infrasight there are no functional requirements, and I'm thinking that keeping the > guards in place will allow for us to try to make lots of requests that fail but then suddenly succeed. I'm not even sure that will help the Infrasight test, but seems like it might.

As for the commented out scheduling calls, I will add comments to them saying that we want to get to them eventually because automatic starting and ending of events will eventually be needed, but not for a while

**Review by dag-yoppworks — APPROVED** [review]
Alright, looking good, thanks.
